src/session.rs

```diff
@@ -36,100 +48,261 @@ pub struct Session<FS: Filesystem> {
+                    .await
+                {
+                    let filesystem = filesystem.clone();
+                    let sender = sender.clone();
+
+                    // tokio::task::spawn(async move {
+                    match req.dispatch(filesystem, sender).await {
+                        Ok(_) => {}
+                        Err(e) => {
+                            warn!("I/O failure in dispatch paths: {:#?}", e);
+                        }
+                    };
+                    buffer = req.data;
+                // });
+                } else {
+                    buffer = Vec::with_capacity(BUFFER_SIZE);
+                }
+            } else {
+                buffer = Vec::with_capacity(BUFFER_SIZE);
             }
         }
-        Ok(())
     }
-}
 
-impl<FS: 'static + Filesystem + Send> Session<FS> {
-    /// Run the session loop in a background thread
-    pub fn spawn(self) -> io::Result<BackgroundSession> {
-        BackgroundSession::new(self)
+    pub(crate) async fn spawn_worker_loop(
+        active_session: Arc<ActiveSession>,
+        ch: Arc<AsyncFd<channel::FileDescriptorRawHandle>>,
+        filesystem: Arc<FS>,
+        worker_idx: usize,
+    ) -> io::Result<()> {
+        let sender = ChannelSender { fd: ch.clone() };
+        loop {
+            let buffer: Vec<u8> = Vec::with_capacity(BUFFER_SIZE);
+
+            if active_session
+                .destroyed
+                .load(std::sync::atomic::Ordering::Relaxed)
+            {
+                return Ok(());
+            }
+
+            if let Some(req_or_err) = Session::<FS>::read_single_request(&ch, buffer).await {
+                let req = req_or_err?;
+                if !active_session
+                    .initialized
+                    .load(std::sync::atomic::Ordering::Relaxed)
+                {
+                    req.dispatch_init(&active_session, &filesystem, sender.clone())
+                        .await;
+                } else {
+                    if !req
+                        .maybe_destroy_dispatch(&active_session, sender.clone())
+                        .await
+                    {
+                        let filesystem = filesystem.clone();
+                        let sender = sender.clone();
+
+                        match req.dispatch(filesystem, sender).await {
+                            Ok(_) => {}
+                            Err(e) => {
+                                warn!("I/O failure in dispatch paths: {:#?}", e);
+                            }
+                        };
+                    }
+                }
+
+                if active_session
+                    .initialized
+                    .load(std::sync::atomic::Ordering::Relaxed)
+                {
+                    break;
+                }
+            }
+        }
+
+        Session::main_request_loop(&active_session, &ch, &filesystem, worker_idx).await
+    }
+
+    async fn driver_evt_loop(
```

Comment:
Why do we need this function? If I understand correctly, it checks every 5secs to see if the Destroy message has been received, and if it has then it destroys the Filesystem. However, can't we call `Filesystem.destroy()` in `maybe_destory_dispatch()`? It's accessible in the only two places that function is called

Reply:
it came into one of the previous comments too -- but this is related to the ownership/shared ownership of the Arc/mutability. The drive event loop will try wait till all the children are shut down so it can take mutable ownership of the contents of the Arc back, and then call the mutable destroy method. This also ensures we avoid any possible races where one channel is doing a destroy and another channel is processing a message before that, which could do a destroy concurrently with an op. 

However not wed to this at all. If we said destroy had a looser contract here around `&self` and maybe you could have a race and deal with it in your filesystem that would work too and simplify this?

Reply:
Ah yes, I see. How about moving the logic that acquires the mutable Filesystem into `dispatch`? I think after setting `destroy.store(true)` it could loop until it can acquire a mutable reference and then call `destroy()`

Reply:
That seems fine, my only hesitance here is that it can be a bit of a foot gun in that, now the outer  section gets to ‘regain control’, where as if the outer loop keeps a reference to the Arc/file system at all now it would hang. But i think if we got that wrong it would reliably hang ? So might be worth just doing the less code/packing it deeper in there. Happy to make that change

Reply:
I tried to do this in dispatch, but it winds up being pretty horrible pretty fast, since ownership/mutability needs to get passed around. It means each Arc can only be really cloned once into each worker and the outside driver not keep a copy. Then a mutable ref of the Arc passed everywhere. Its doable, but its pretty ugly/awkward looking. And would be very easy to accidentally break that i can see. One `Arc::clone()` in any path inside a worker loop would make it fail

Reply:
Ok, lets stick with the driver loop then :) Thanks for looking into it!